### PR TITLE
Add shape comment for some tensors and remove 2 'for' loops

### PR DESCRIPTION
**Pull request by betterenvi**
## What this PR does:
- Add shape comment for some tensors and add some docstrings for some methods

These comments make the program more readable.

- Remove the for-loop in `calculate_new_cell_state` and `calculate_new_hidden_state` of `HMLSTMCell`

- Fix docstring for `train` method of `HMLSTMNetwork`

In the `train` method of `HMLSTMNetwork`, `batch_in`'s shape should be `[num_batches, batch_size, num_timesteps, input_size]` instead of `[num_batches, num_timesteps, batch_size, input_size]`

### CODE REVIEW

**Comment by nsfinkelstein** [comment]
@betterenvi 

Thanks for making another pull request!  I like the idea of the tensor shape comments.

Can you add some more comments indicating where the 'flush', 'update', and 'copy' operations are taking place where you've replaced the for loop? I think that would make it easier to follow along with the paper.

Also, just to check, have you tested that this change works exactly as you'd expect? I'm not very familiar with `tf.where`.

**Comment by betterenvi** [comment]
I added some more comments indicating where the 'flush', 'update', and 'copy' operations are taking place.

As for `tf.where`, I run a language model before and after this change, and found that the loss was decreasing at almost same speed (measured by num of steps).

**Comment by nsfinkelstein** [comment]
@betterenvi 

Thanks. 

I'd want to do some more testing on the functions that update `c` and `h` - I'd expect loss to decrease at roughly the same rate whether the model was coded correctly or whether there were certain kinds of bugs (e.g. if c and h got updated at every step).  

That part was a little tricky to get right, so I don't want to replace the code without doing an equivalent amount of testing as I did the first time around.  If you can do some further testing that proves those functions maintain they're behavior (ideally automated tests), that would be great.

Otherwise, if you separate out the commits adding comments into they're own pull request, I can merge that now, and then I'll merge the rest of it when I have time to do the testing myself. (That may not be for some time).

**Comment by betterenvi** [comment]
@n-s-f 

Thanks for your time!

I will do more tests ASAP.

**Comment by betterenvi** [comment]
I dit a test.

Test code: test.py. Set random seed before generating data and training
```Python
import time
import numpy as np
import tensorflow as tf
from hmlstm import (
    HMLSTMNetwork, plot_indicators, prepare_inputs,
    get_text, convert_to_batches, viz_char_boundaries
)

tf.set_random_seed(1)
np.random.seed(1)

num_signals = 300
signal_length = 400
x = np.linspace(0, 50 * np.pi, signal_length)
signals = [np.random.normal(0, .5, size=signal_length) +
           (2 * np.sin(.6 * x + np.random.random() * 10)) +
           (5 * np.sin(.1* x + np.random.random() * 10))
    for _ in range(num_signals)]

print(sum([s.sum() for s in signals]))

split = int(num_signals * .8)
train = signals[:split]
test = signals[split:]

batches_in, batches_out = convert_to_batches(signals, batch_size=10)

st = time.time()
network = HMLSTMNetwork(input_size=1, task='regression', hidden_state_sizes=30,
                       embed_size=50, out_hidden_size=30, num_layers=2)

network.train(batches_in[:-1], batches_out[:-1], epochs=1,
              load_vars_from_disk=False,
              save_vars_to_disk=False,
              variable_path='./sinusoidal')

ed = time.time()
print('Time:', ed - st)
```

log:
```
$ CUDA_VISIBLE_DEVICES=0  python test.py 
-800.109262628
2017-07-26 13:20:21.379430: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:20:21.379457: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:20:21.379462: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:20:21.379465: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:20:21.379469: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:20:21.659317: I tensorflow/core/common_runtime/gpu/gpu_device.cc:940] Found device 0 with properties: 
name: GeForce GTX TITAN X
major: 5 minor: 2 memoryClockRate (GHz) 1.076
pciBusID 0000:0a:00.0
Total memory: 11.92GiB
Free memory: 11.81GiB
2017-07-26 13:20:21.659345: I tensorflow/core/common_runtime/gpu/gpu_device.cc:961] DMA: 0 
2017-07-26 13:20:21.659350: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   Y 
2017-07-26 13:20:21.659357: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:0a:00.0)
Epoch 0
2017-07-26 13:20:28.190616: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 104081 get requests, put_count=102573 evicted_count=1000 eviction_rate=0.00974915 and unsatisfied allocation rate=0.0250574
2017-07-26 13:20:28.190651: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 100 to 110
loss: 19.754
loss: 19.3416
loss: 19.0624
loss: 18.0714
2017-07-26 13:20:54.061321: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 195873 get requests, put_count=195960 evicted_count=2000 eviction_rate=0.0102062 and unsatisfied allocation rate=0.00985332
2017-07-26 13:20:54.061363: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 193 to 212
loss: 17.9569
loss: 18.145
loss: 17.0898
loss: 16.9008
loss: 17.6258
2017-07-26 13:21:29.278319: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 195773 get requests, put_count=195952 evicted_count=2000 eviction_rate=0.0102066 and unsatisfied allocation rate=0.00944461
2017-07-26 13:21:29.278361: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 309 to 339
loss: 15.7292
loss: 15.657
loss: 16.1844
loss: 15.9067
loss: 15.8196
loss: 15.5662
2017-07-26 13:22:09.199397: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 207121 get requests, put_count=207591 evicted_count=2000 eviction_rate=0.00963433 and unsatisfied allocation rate=0.00762356
2017-07-26 13:22:09.199432: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 542 to 596
loss: 13.7765
loss: 13.8101
loss: 14.2227
loss: 13.5495
loss: 13.5155
loss: 12.5116
loss: 12.4024
loss: 12.4842
2017-07-26 13:23:02.097072: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 195282 get requests, put_count=195424 evicted_count=1000 eviction_rate=0.00511708 and unsatisfied allocation rate=0.00493133
2017-07-26 13:23:02.097102: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 1158 to 1273
loss: 11.9458
loss: 11.3283
loss: 10.9446
loss: 10.1651
loss: 9.96376
saving variables...
Time: 195.15480732917786
```

And then, I replace the `hmlstm_cell.py` with my version

```
$ cd  hmlstm/
$ cp my_hmlstm_cell.py hmlstm_cell.py 
$ cd  ..
$ CUDA_VISIBLE_DEVICES=0  python test.py 
-800.109262628
2017-07-26 13:24:23.546418: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:24:23.546445: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:24:23.546451: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:24:23.546457: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:24:23.546462: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-07-26 13:24:23.882050: I tensorflow/core/common_runtime/gpu/gpu_device.cc:940] Found device 0 with properties: 
name: GeForce GTX TITAN X
major: 5 minor: 2 memoryClockRate (GHz) 1.076
pciBusID 0000:0a:00.0
Total memory: 11.92GiB
Free memory: 11.81GiB
2017-07-26 13:24:23.882081: I tensorflow/core/common_runtime/gpu/gpu_device.cc:961] DMA: 0 
2017-07-26 13:24:23.882088: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   Y 
2017-07-26 13:24:23.882097: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:0a:00.0)
Epoch 0
2017-07-26 13:24:25.679763: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 34442 get requests, put_count=34250 evicted_count=1000 eviction_rate=0.0291971 and unsatisfied allocation rate=0.0375123
2017-07-26 13:24:25.679798: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 100 to 110
loss: 19.7981
loss: 19.3877
loss: 19.1105
loss: 18.1178
loss: 18.0083
loss: 18.2129
loss: 17.1666
loss: 16.9967
loss: 17.7594
2017-07-26 13:24:36.339967: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 34448 get requests, put_count=34410 evicted_count=1000 eviction_rate=0.0290613 and unsatisfied allocation rate=0.030742
2017-07-26 13:24:36.340004: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 233 to 256
loss: 15.8584
loss: 15.8098
loss: 16.3762
loss: 16.1258
loss: 16.0769
loss: 15.8654
loss: 14.0959
loss: 14.1772
loss: 14.6569
loss: 14.0093
loss: 14.0155
loss: 13.037
2017-07-26 13:24:50.875503: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 67309 get requests, put_count=66923 evicted_count=1000 eviction_rate=0.0149425 and unsatisfied allocation rate=0.0213939
2017-07-26 13:24:50.875545: I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 596 to 655
loss: 12.9701
loss: 13.1133
loss: 12.6278
loss: 12.0482
loss: 11.7274
loss: 10.9998
loss: 10.9321
saving variables...
Time: 35.88774609565735
```

As the logs indicated, the loss was decreasing at almost same speed in terms of num of steps, before and after my change. But the time was significantly reduced after the change.

The reason why the loss was not absolutely the same was that the graph had been changed. I did a test, and found that only adding one following line will have a different result.
``` Python
z = z + 0  # add this line to calculate_new_cell_state in HMLSTMCell
```

**Comment by ligaoheng** [comment]
Yes, the `[]` operation is very time-consuming!

**Comment by nsfinkelstein** [comment]
@betterenvi Thanks for doing this test! I'm glad to see that this change will speed up the training.

Can you do one other test? 

Can you put in some save ops to make sure that the hidden and cell states don't change when they're not supposed to?

For example, you can use [tf.summary.scalar](https://www.tensorflow.org/api_docs/python/tf/summary/scalar) to save the value of the relevant indicator variables, and to save the means of the hidden and cell states, for every batch.  

Then you could look to see if the indicators are both zero, are the means of the hidden and cell states exactly the same as in the previous time step for that batch? Can each batch behave differently, as it's supposed to?

If you see that this works as expected, I'll merge. I wouldn't expect a big difference in the way the loss decreases between, for example, this model and a simple stacked LSTM, which is what this model might effectively become if that piece of code does not work as expected.

**Comment by betterenvi** [comment]
@n-s-f 

Sorry, I don't catch your point. 

What do mean by **`the hidden and cell states don't change when they're not supposed to`**?  We don't know what the states are supposed to be. If you mean: we can conclude that this change works expected by comparing the states' values before and after my change. But I think that's impossible. As I did the test before, even if I set same random seed and only added one line `z = z + 0`, the initialization of variables would change, since the graph had been changed.


Think about what this change does. I replaced the 2 for loops with [`tf.where`](https://www.tensorflow.org/api_docs/python/tf/where). As the doc says, `tf.where` **“Return the elements, either from x or y, depending on the condition.”**, which is exactly what the original 2 loops do. That is to say, `tf.where` and original for loop are logically identical. 

Therefore, I don't think we need another test.

**Comment by nsfinkelstein** [comment]
@betterenvi 

Thanks again for your work on this

I agreed that we don't know what the states are supposed to be, but we do know when they're supposed to change from one time step to the next based on the indicators, so we can tell if they're changing when they're not supposed to change.  This can be determined in the manner I described.  It's also possible to use the [tensorflow debugger](https://www.tensorflow.org/programmers_guide/debugger) to make sure the calculations are being carried through as expected, which is what I did for the current code.  I'll try to get to doing one of these kinds of tests this weekend.

I also agree that it seems like these pieces of code should be equivalent. However, I've had a number of experiences where logically I expect tensorflow to behave one way, and it behaves a different way :)  Of note, the `case` and `cond` functions accept functions as arguments, the `where` function accepts tensors, and I'm not really sure why that is or what affect that has on the computational graph.

**Comment by betterenvi** [comment]
@n-s-f 

I see it now. 

I'm not familiar with tensorflow debugger, and it would be great if you can share your test code.

Thanks!

**Comment by nsfinkelstein** [comment]
Looks good :+1:
